### DISCUSSION TITLE
Decode extension object in Android application

### DISCUSSION THREAD

**Discussion by sergeishugaev** [eclipse-milo/milo#1319] [General]
Hello!

I am writing an application using the Milo library for Android. 
I need to get an array of structures from the PLC.
I have written two programs: the first program is a simple Java program, the second is a similar java program for Android. 
The first program works successfully, the second throws an exception at the decoding stage of the Extension Object.
`Object object = extensionObject.decode(opcUaClient.getDynamicSerializationContext());`
Exception:
`org.eclipse.milo.opcua.stack.core.UaSerializationException: no codec registered for encodingId=NodeId{ns=4, id=|enc|SPK107 (M01).Application.SHIPMENT_AUTO}`

Please tell me the direction of finding a solution to this problem.

Perhaps the Android application needs some additional permissions. Tell me, when decoding, what resources does the milo library use?

Thanks!

**Comment by sergeishugaev** [comment]
UAExpert
![image](https://github.com/user-attachments/assets/132ffa1c-18fc-45b4-9f2f-a274d4f4236f)

**Comment by kevinherron** [comment]
It sounds like reading the DataType dictionary failed. Maybe there was an error message?

**Comment by Unchakable** [comment]
I also faced this problem. If I understood correctly, BsdParser uses the awt package, but it is not in the sdk. There is no way to add it. Is there any way to do without BsdParser when developing mobile applications for android?

Rejecting re-init on previously-failed class java.lang.Class<com.sun.xml.bind.v2.model.impl.RuntimeBuiltinLeafInfoImpl>: java.lang.NoClassDefFoundError: Failed resolution of: Ljava/awt/Image;

at org.eclipse.milo.opcua.binaryschema.parser.DictionaryDescription org.eclipse.milo.opcua.binaryschema.parser.BsdParser.parse(java.io.InputStream) (BsdParser.java:40)

**Reply by kevinherron** [reply]
Hmm, it's not using it directly, but it looks some Java XML implementation is...

You can try using the `dev/1.0` branch which has support for the DataType Definition attribute, and uses that by default instead of reading the legacy XML-based DataType Dictionary.

See these examples:
- https://github.com/eclipse/milo/blob/dev/1.0/milo-examples/client-examples/src/main/java/org/eclipse/milo/examples/client/UnifiedAutomationReadCustomDataTypeExample1.java
- https://github.com/eclipse/milo/blob/dev/1.0/milo-examples/client-examples/src/main/java/org/eclipse/milo/examples/client/UnifiedAutomationReadCustomDataTypeExample2.java

**Reply by Unchakable** [reply]
Unfortunately, this did not solve the problem. Other errors occur. I thought I could implement some of the functionality myself for my little task. Now I'm getting a valid ExtensionObject. I can get a ByteString list from it. If I understood correctly, these are the values of variables in the form of an array of bytes. Can I get a sequential list of variables from the controller to assign them a value from the ByteString?

**Reply by kevinherron** [reply]
>Can I get a sequential list of variables from the controller to assign them a value from the ByteString?

No, that's not what an ExtensionObject is. An ExtensionObject is a container for an encoded structure - you need to decode it into a structure.

Feel free to start a new discussion where you share actual code and error messages.

**Reply by sergeishugaev** [reply]
I've done some tests and I think the problem is not with the parser, but with OpcUaClient.
I test simple java app and android app and I noticed that on android there is no corresponding Extension object codec (opcUaClient - > stackClient -> codecsByEncodingId) in the opcUaClient after execute opcUaClient.connect().get()
What do you think about it?

**Reply by kevinherron** [reply]
> opcUaClient.connect().get()

This implies you are back to using version 0.6.x?

In this case, you must add a SessionInitializer that will attempt to read the datatype dictionary after a Session is established. See https://github.com/eclipse/milo/blob/d73a2165b9b3cf9f9cb0dd8311c19e2adab03863/milo-examples/client-examples/src/main/java/org/eclipse/milo/examples/client/UnifiedAutomationReadCustomDataTypeExample.java#L43-L48

There will be no codecs if this isn't done, or if reading the dictionary fails for some reason.

**Reply by sergeishugaev** [reply]
I probably put it wrong. The Android application does not receive dictionaries, and therefore codecs. As I understand it, dictionaries and codecs are received when executing opcUaClient.connect().get().
The java app:
![2024-10-03_15-43-06](https://github.com/user-attachments/assets/aaed9ef8-d6aa-4812-9d45-715a2547a080)
The same android app:
![2024-10-03_15-43-57](https://github.com/user-attachments/assets/dce2d4ce-72f8-40c2-b016-584949cb0c0d)

I would like to get the dictionary from the plc and not define the custom data type.

**Reply by kevinherron** [reply]
You need that line I mentioned above that reads the datatype dictionary. If you already have that and still have no dictionaries, there is some error you need to track down.

**Reply by sergeishugaev** [reply]
How do I get the dictionaries again, after opcUaClient.connect().get(), is there any method?
How can I track down the error, give me some advice.

**Reply by kevinherron** [reply]
@Unchakable is reporting there is some error due to a missing awt package that prevents this from working.

> Rejecting re-init on previously-failed class java.lang.Class<com.sun.xml.bind.v2.model.impl.RuntimeBuiltinLeafInfoImpl>: java.lang.NoClassDefFoundError: Failed resolution of: Ljava/awt/Image;
> 
> at org.eclipse.milo.opcua.binaryschema.parser.DictionaryDescription org.eclipse.milo.opcua.binaryschema.parser.BsdParser.parse(java.io.InputStream) (BsdParser.java:40)

**Reply by kevinherron** [reply]
`DataTypeDictionaryReader` has a logger that logs warnings, and would log additional useful information if you set it to DEBUG level. 

I don't know how this is done on Android, or even where your log output goes to.

**Comment by kevinherron** [comment]
From what I can gather via Googlin', JAXB is not supported on Android, and JAXB is required for reading the DataType Dictionary, so you won't be able to use 0.6.x on Android if you need to read custom types.

In the `dev/1.0` branch you can rely on the DataTypeDefinition attribute instead, which doesn't use XML/JAXB at all. If there are other Android problems lurking you'll have to find and report them, preferably in a new discussion.
